src/builtin.cpp

```diff
@@ -210,42 +210,6 @@ static maybe_t<int> implemented_in_rust(parser_t &, io_streams_t &, const wchar_
     DIE("builtin is implemented in Rust, this should not be called");
 }
 
-// How many bytes we read() at once.
-// Since this is just for counting, it can be massive.
-#define COUNT_CHUNK_SIZE (512 * 256)
-/// Implementation of the builtin count command, used to count the number of arguments sent to it.
-static maybe_t<int> builtin_count(parser_t &parser, io_streams_t &streams, const wchar_t **argv) {
-    UNUSED(parser);
-    int argc = 0;
-
-    // Count the newlines coming in via stdin like `wc -l`.
-    if (streams.stdin_is_directly_redirected) {
-        assert(streams.stdin_fd >= 0 &&
-               "Should have a valid fd since stdin is directly redirected");
-        char buf[COUNT_CHUNK_SIZE];
-        while (true) {
-            long n = read_blocked(streams.stdin_fd, buf, COUNT_CHUNK_SIZE);
-            if (n == 0) {
-                break;
-            } else if (n < 0) {
-                wperror(L"read");
-                return STATUS_CMD_ERROR;
```

Comment:
This errored out completely if `read` failed, and printed an error message. Arguments just silently ignores those errors. We can either duplicate code from Arguments directly here or add:

```rust
pub struct ResultArguments(Arguments);
```

And make `get_arg_stdin` return an `io::Result`, forward that directly in `Iteraror for ResultArguments`, while moving the `.ok()` to `next()` in `Iteraror for Arguments`

We could also add `StdinOnlyArguments` or another constructor to `ResultArguments` that makes it unnecessary to send the dummy-`args`

Reply:
Note that you can `let res: io::Result<usize> = iter_of_result.collect()`

Reply:
> This errored out completely if read failed, and printed an error message. Arguments just silently ignores those errors.

This is the biggest change here, but tbh I've looked over how `read` *can* fail - assuming this fails like read(2), and I am not sure what the failure modes here *are*?

My man page says:

- EBADF, EFAULT, EINVAL, EISDIR - these are checked when we open the fd
- EAGAIN / EWOULDBLOCK - only relevant for O_NONBLOCK, which I think we would have to do on the fd, and we don't?
- EIO - *possibly* for network filesystems when the lock is lost? maybe?
- EINTR - handled by read_blocked
- "other errors"

> Note that you can let res: io::Result<usize> = iter_of_result.collect()


Sorry, I don't understand what you are saying.

Reply:
> We could also add StdinOnlyArguments or another constructor to ResultArguments that makes it unnecessary to send the dummy-args

I said as much in the opening message. My gut says this isn't a problem often enough to warrant it.

Reply:
>  > Note that you can let res: io::Result = iter_of_result.collect()

> Sorry, I don't understand what you are saying.

I thought it might be relevant to mention that you can short-circuit if you have an iterator of `Result`, which is not a super-obvious feature of `.collect()`. I wrongly used  `io::Result<usize>` instead of `io::Result<Vec<bool>>` which I think caused the confusion, as you cannot collect to a `usize`

```rust
fn main() {
    let results = [Ok(1), Ok(2), Err("Failed"), Ok(4)];
    let filtered: Result<Vec<usize>, &str> = results.into_iter().collect();
    // prints Err("Failed")
    println!("{:?}", filtered);
    let results = [Ok(1), Ok(2), Ok(3), Ok(4)];
    let filtered: Result<Vec<usize>, &str> = results.into_iter().collect();
    // prints Ok([1, 2, 3, 4])
    println!("{:?}", filtered);
}
```

Reply:
Thinking over it, given that we ignore the result in string and path, I agree with you that this can just be ignored, would be weird to handle it here but not there

Reply:
Funnily enough I had a comment to that effect originally but removed it for being weird meta-commentary that I tend to add too much of.